### PR TITLE
Poseidon BB machine.

### PR DESCRIPTION
**Pull request by lvella**
Working with small memory machine.

### CODE REVIEW

**Comment by lvella** [comment]
This time targeting main.

File: test_data/std/poseidon_gl_memory_test.asm [hunk 1]
```diff
@@ -137,7 +137,7 @@ machine Main with degree: 65536 {
         mstore_le 180, 0, 0;
         mstore_le 188, 0, 0;
  
-        // This will read bytes [100, 191] and write the result to bytes [104, 131]
+        // This will read bytes [100, 195] and write the result to bytes [104, 131]
```
Comment (lvella) [inline test_data/std/poseidon_gl_memory_test.asm:140]: I think this line is yours, @leonardoalt . Is it correct? Should be left here?

Comment (leonardoalt) [reply test_data/std/poseidon_gl_memory_test.asm:140]: it came from your original poseidon BB PR I think?

Comment (lvella) [reply test_data/std/poseidon_gl_memory_test.asm:140]: I remember now. I fixed the value when porting the test to poseidon-bb. I re-ran the numbers and it is correct.

**Comment by leonardoalt** [comment]
Linter is failing. I guess that's why @Schaeff added https://github.com/powdr-labs/powdr/pull/1925/files#diff-e9cc0c11ed0e05388d68419f06b32c0bb9844f29dcde2fa503fd79ace4e32c49R2 here.

**Comment by lvella** [comment]
how this got into main?

**Comment by leonardoalt** [comment]
> how this got into main?

also no idea

File: std/machines/hash/poseidon_bb.asm [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+    // Increment the low limb if address is not being set:
+    (1 - is_addr_set') * (
+        // If low limb is about to overflow, next value must be 0:
+        low_overflow * low_addr' +
+        // Otherwise, next value is current plus 4:
+        (1 - low_overflow) * (low_addr + 4 - low_addr')
+    ) = 0;
+    // Set high limb, incremented if low overflowed:
+    (1 - is_addr_set') * (high_addr' - high_addr - low_overflow) = 0;
```
Comment (leonardoalt) [inline std/machines/hash/poseidon_bb.asm:120]: Does this block assume addresses to be 4-aligned?

**Comment by leonardoalt** [comment]
Lint still failing. There are a few more that the other PR also fixes.

File: std/machines/hash/poseidon_bb.asm (hunk 2 above)
Comment (lvella) [reply std/machines/hash/poseidon_bb.asm:120]: Yes. Starting from this line:
```
    let low_diff = low_addr - (0x10000 - 4);
```

 My hope is that address alignment is enforced somewhere else (in memory machine?). I know it is enforced for poseidon_gl when the call comes from RISC-V.

Comment (leonardoalt) [reply std/machines/hash/poseidon_bb.asm:120]: yea I think it makes sense

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**